experiments/tree_tensor_network/ttn_1d_uniform.py

```diff
@@ -879,64 +1739,562 @@ def descend_full_state_pure(isos_012):
+  """Creates a 'blocked' Hamiltonian from an input Hamiltonian.
+
+  Blocks sites together, increasing the site dimension.
+
+  Args:
+    H: The local Hamiltonian term.
+    sites_per_block: The number of sites to block into one.
+
+  Returns:
+    The blocked local Hamiltonian term.
+  """
+  h1, h2 = H
+  d = h1.shape[0]
+  dtype = h1.dtype
+  E = backend.eye(d, dtype=dtype)
+
+  h1_blk = None
+  for i in range(sites_per_block):
+    h1_term = h1 if i == 0 else E
+    for j in range(1, sites_per_block):
+      h1_term = kron_td(h1_term, h1 if i == j else E)
+    if h1_blk is not None:
+      h1_blk += h1_term
+    else:
+      h1_blk = h1_term
+
+  h2_dense = sum(kron_td(h2[0][i], h2[1][i]) for i in range(len(h2[0])))
+  for i in range(sites_per_block - 1):
+    h1_term = h2_dense if i == 0 else E
+    j = 2 if i == 0 else 1
+    while j < sites_per_block:
+      h1_term = kron_td(h1_term, h2_dense if i == j else E)
+      j += 2 if i == j else 1
+    h1_blk += h1_term
+  del(h2_dense)
+
+  E_big = backend.eye(d**(sites_per_block - 1), dtype=dtype)
+  h2_0 = [kron_td(E_big, h) for h in h2[0]]
+  h2_1 = [kron_td(h, E_big) for h in h2[1]]
+
+  return h1_blk, (h2_0, h2_1)
+
+
 def get_ham_ising_tube(dtype, Ly, lam=-3.044):
-  X = np.array([[0.0, 1.0], [1.0, 0.0]])
-  Z = np.array([[1.0, 0.0], [0.0, -1.0]])
+  """Return the local term for the 2+1D Ising Hamiltonian on a narrow torus.
+
+  Defines the global Hamiltonian:
+  $H = -\sum_{\langle i, j \rangle} X_i X_j + lam * \sum_i Z_i ]$
+
+  Represents the Hamiltonian for the 2D torus as a 1-dimensional Hamiltonian,
+  where each "site" is a slice of the torus in the "y" direction. The site
+  dimension thus depends on the size of the system in the y direction.
+
+  Args:
+    dtype: The data type.
+    Ly: The size of the torus in the y direction (number of sites).
+    lam: The field strength.
+
+  Returns:
+    The Hamiltonian term, separated into a 1-site contribution and a 2-site
+    MPO.
+  """
+  X = backend.np.array([[0.0, 1.0], [1.0, 0.0]])
+  Z = backend.np.array([[1.0, 0.0], [0.0, -1.0]])
 
   Xcol = [
-      np.kron(np.kron(np.eye(2**i), X), np.eye(2**(Ly - i - 1)))
+      backend.np.kron(
+        backend.np.kron(backend.np.eye(2**i), X),
+        backend.np.eye(2**(Ly - i - 1)))
       for i in range(Ly)
   ]
   Zcol = [
-      np.kron(np.kron(np.eye(2**i), Z), np.eye(2**(Ly - i - 1)))
+      backend.np.kron(
+        backend.np.kron(backend.np.eye(2**i), Z),
+        backend.np.eye(2**(Ly - i - 1)))
       for i in range(Ly)
   ]
 
-  Xcol = [tf.convert_to_tensor(Xc, dtype=dtype) for Xc in Xcol]
-  Zcol = [tf.convert_to_tensor(Zc, dtype=dtype) for Zc in Zcol]
+  Xcol = [backend.convert_to_tensor(Xc, dtype=dtype) for Xc in Xcol]
+  Zcol = [backend.convert_to_tensor(Zc, dtype=dtype) for Zc in Zcol]
 
   h1 = lam * sum(Zcol) - sum(Xcol[i] @ Xcol[(i + 1) % Ly] for i in range(Ly))
 
   h_mpo_2site = ([-Xc for Xc in Xcol], Xcol)
 
   return h1, h_mpo_2site
+
+
+class TTNBackend():
```

Comment:
Would it make sense to add these operations to the BaseBacked code in tensornetwork proper? (To be done in a later PR.)

Reply:
Yeah, I wondered about that in a comment above. Do we want to support a more comprehensive API for projects that use TensorNetwork? This would be very useful, but could be tricky due to differing behaviors of different frameworks. I have done the minimal necessary implementation for the tree code here, but it would need upgrading to make it more general.